File: qeli-shared/QeliShared/Protocol/PacketCodec.cs
using System.Security.Cryptography;
using Qeli.Shared.Crypto;

namespace Qeli.Shared.Protocol;

public sealed class PacketException : Exception
{
    public PacketException(string message) : base(message) { }
}

/// <summary>
/// Frames/deframes data-plane records. Direct port of Android PacketCodec.kt.
/// Wire layout: TLS record header [0x17 0x03 0x03 len_hi len_lo] || nonce(12) ||
/// ChaCha20-Poly1305( counter(8) || plaintext || padding || pad_len(2) ).
/// Includes the same 64-entry anti-replay sliding window as the server.
/// </summary>
public sealed class PacketCodec
{
    public const int HeaderSize = 5;
    public const int NonceSize = 12;
    public const int TagSize = 16;
    public const int CounterSize = 8;
    public const int ReplayWindow = 2048; // WireGuard-sized anti-replay window (M-13)
    public const int ReplayWords = ReplayWindow / 64;
    public const byte ApplicationData = 0x17;
    public const int MaxRecordSize = 16384 + NonceSize + TagSize + CounterSize + 256;

    private readonly PacketCipher _cipher;
    private bool _paddingEnabled;
    private int _paddingMin;
    private int _paddingMax;

    // Wire framing. TLS (5-byte 0x17 0x03 0x03 + u16 len) for fake-tls/obfs/reality;
    // Raw (bare u16 len, RAW_RECORD_HEADER) for the `plain` wire mode. Mirrors the
    // Rust PacketCodec Framing::Tls / Framing::Raw.
    private readonly bool _raw;
    private readonly int _headerSize;

    private long _counter;            // outbound, monotonically increasing
    // "Not initialised yet" used to be encoded as `_replayHighest < 0`, which collides with
    // every counter whose top bit is set. One record with a counter >= 2^63 — the sequence
    // comes straight out of the decrypted plaintext, so a hostile or compromised server picks
    // it — left _replayHighest negative, and from then on the `< 0` branch fired on EVERY
    // packet and returned true unconditionally: the window was off for the rest of the
    // session and any captured record could be replayed at will. Rust keeps a separate
    // `initialized` flag and a u64 counter; Swift uses `UInt64?`. Only C# and Kotlin encoded
    // the sentinel in-band. (Audit 2026-08-04, H-06.)
    private bool _replayInitialized;
    private long _replayHighest;      // inbound replay window (unsigned value in a long)
    private readonly ulong[] _replayBits = new ulong[ReplayWords]; // 2048-bit window (M-13)

    // M6: per-instance nonce seed + PRP key. The nonce goes on the wire and the peer never
    // inverts the PRP (it reads the nonce off the wire), so these are local randomness and
    // need NOT match the peer's — they only have to make (seed‖counter) unique per key, which
    // a monotonic counter + fresh per-session key guarantee. Fresh seed per instance also
    // keeps nonces unique across a reconnect that reused the key.
    private readonly byte[] _nonceSeed = RandomNumberGenerator.GetBytes(4);
    private readonly byte[] _noncePrpKey = RandomNumberGenerator.GetBytes(32);

    public PacketCodec(PacketCipher cipher, bool paddingEnabled = true, int paddingMin = 0, int paddingMax = 255,
        bool raw = false)
    {
        _cipher = cipher;
        _paddingEnabled = paddingEnabled;
        _paddingMin = paddingMin;
        _paddingMax = paddingMax;
        _raw = raw;
        _headerSize = raw ? 2 : HeaderSize;
    }

    /// <summary>Apply server-pushed padding params without resetting the packet counter.</summary>
    public void SetPadding(bool enabled, int min, int max)
    {
        _paddingEnabled = enabled;
        _paddingMin = min;
        _paddingMax = max;
    }

    /// <summary>Anti-replay verdict for <paramref name="seq"/>, recording it as seen.
    /// `internal` so the shared replay-window fixture (<c>conformance/replay-window.json</c>)
    /// can drive it directly — the window is pure state, and going through Decrypt would
    /// need a valid record per sequence number.</summary>
    internal bool AcceptCounter(long seq)
    {
        // The counter is an UNSIGNED 64-bit wire value; `long` only holds its bit pattern,
        // so every comparison below goes through `ulong`. See _replayInitialized.
        ulong s = (ulong)seq, highest = (ulong)_replayHighest;
        if (!_replayInitialized)
        {
            _replayInitialized = true;
            _replayHighest = seq;
            _replayBits[0] = 1UL;
            return true;
        }
        if (s > highest)
        {
            ulong advance = s - highest;
            if (advance >= ReplayWindow) Array.Clear(_replayBits, 0, ReplayWords);
            else ShiftWindow((int)advance);
            _replayHighest = seq;
            _replayBits[0] |= 1UL; // distance 0 = current highest seq
            return true;
        }
        ulong diff = highest - s;
        if (diff >= ReplayWindow) return false;
        ulong mask = 1UL << (int)(diff % 64);
        int wi = (int)(diff / 64);
        if ((_replayBits[wi] & mask) != 0) return false;
        _replayBits[wi] |= mask;
        return true;
    }

    /// <summary>Multi-word left shift of the replay window by <paramref name="n"/> bits
    /// (toward higher distance), discarding bits that fall off the top.</summary>
    private void ShiftWindow(int n)
    {
        int words = n / 64, off = n % 64;
        if (off == 0)
            for (int i = ReplayWords - 1; i >= 0; i--)
                _replayBits[i] = i >= words ? _replayBits[i - words] : 0UL;
        else
            for (int i = ReplayWords - 1; i >= 0; i--)
            {
                ulong lo = i >= words ? _replayBits[i - words] << off : 0UL;
                ulong hi = i > words ? _replayBits[i - words - 1] >> (64 - off) : 0UL;
                _replayBits[i] = lo | hi;
            }
    }

    private static byte[] BuildTlsRecordHeader(byte contentType, int length) => new[]
    {
        contentType, (byte)0x03, (byte)0x03,
        (byte)((length >> 8) & 0xFF), (byte)(length & 0xFF),
    };

    public byte[] Encrypt(byte[] plaintext)
    {
        int paddingLen = 0;
        if (_paddingEnabled)
        {
            int lo = Math.Clamp(_paddingMin, 0, 65535);
            int hi = Math.Clamp(_paddingMax, lo, 65535);
            paddingLen = hi > lo ? lo + RandomNumberGenerator.GetInt32(hi - lo + 1) : lo;
        }
        return EncryptPadded(plaintext, paddingLen);
    }

    /// <summary>Encrypt with the configured padding range, but capped so that
    /// plaintext+padding never exceeds <paramref name="maxInnerPlusPad"/>. Keeps the
    /// padded record inside the (probed) tunnel MTU so a DF-marked UDP datagram is not
    /// dropped with EMSGSIZE after path-MTU probing — the server pushes 40–400 B of
    /// padding, which otherwise pushes every full-size data packet past the path MTU.
    /// Mirrors the Rust client's per-packet pad_cap (client/mod.rs).</summary>
    public byte[] EncryptCapped(byte[] plaintext, int maxInnerPlusPad)
    {
        if (!_paddingEnabled) return EncryptPadded(plaintext, 0);
        int room = Math.Max(0, maxInnerPlusPad - plaintext.Length);
        int lo = Math.Clamp(_paddingMin, 0, room);
        int hi = Math.Clamp(_paddingMax, lo, room);
        int pad = hi > lo ? lo + RandomNumberGenerator.GetInt32(hi - lo + 1) : lo;
        return EncryptPadded(plaintext, pad);
    }

    /// <summary>Encrypt with an EXPLICIT padding length, ignoring the codec's
    /// configured padding range. Used by flow-shaping cover traffic to emit
    /// browsing-sized cover packets (empty plaintext + sized padding); the wire
    /// format is byte-identical to <see cref="Encrypt(byte[])"/>.</summary>
    public byte[] EncryptPadded(byte[] plaintext, int paddingLen)
    {
        long currentCounter = Interlocked.Increment(ref _counter) - 1;
        if (currentCounter >= long.MaxValue - 1000)
            throw new PacketException("Counter exhausted - session must be renegotiated");

        // Counter-derived, collision-free, DPI-opaque nonce (was a random 96-bit value,
        // which carries a birthday-bound collision risk the Rust core eliminates). See
        // NonceForCounter / PrpNonce. (client-audit M6)
        var nonce = NonceForCounter(currentCounter);

        paddingLen = Math.Clamp(paddingLen, 0, 65535);
        var padding = new byte[paddingLen];
        if (paddingLen > 0) RandomNumberGenerator.Fill(padding);

        var inner = new byte[CounterSize + plaintext.Length + paddingLen + 2];
        inner[0] = (byte)((currentCounter >> 56) & 0xFF);
        inner[1] = (byte)((currentCounter >> 48) & 0xFF);
        inner[2] = (byte)((currentCounter >> 40) & 0xFF);
        inner[3] = (byte)((currentCounter >> 32) & 0xFF);
        inner[4] = (byte)((currentCounter >> 24) & 0xFF);
        inner[5] = (byte)((currentCounter >> 16) & 0xFF);
        inner[6] = (byte)((currentCounter >> 8) & 0xFF);
        inner[7] = (byte)(currentCounter & 0xFF);
        Buffer.BlockCopy(plaintext, 0, inner, CounterSize, plaintext.Length);
        Buffer.BlockCopy(padding, 0, inner, CounterSize + plaintext.Length, paddingLen);
        inner[^2] = (byte)((paddingLen >> 8) & 0xFF);
        inner[^1] = (byte)(paddingLen & 0xFF);

        var ciphertext = _cipher.Encrypt(inner, nonce);

        int payloadLen = NonceSize + ciphertext.Length;
        // Guard the record size BEFORE writing the 16-bit length field (parity with Rust
        // encrypt_packet's MAX_RECORD_SIZE check). Without it, an oversized padding_max or
        // shaping cover size can build a record the peer rejects as too large (16677-65535),
        // and past 65535 the length write wraps and desyncs the whole TCP stream. Fail here
        // instead of emitting a record we (or the peer) can't parse.
        if (payloadLen > MaxRecordSize)
            throw new PacketException(
                $"record payload {payloadLen} exceeds MaxRecordSize {MaxRecordSize} — " +
                "reduce padding_max or the shaping cover size");

        var packet = new byte[_headerSize + payloadLen];
        if (_raw)
        {
            // Bare 2-byte big-endian length prefix (no TLS type/version).
            packet[0] = (byte)((payloadLen >> 8) & 0xFF);
            packet[1] = (byte)(payloadLen & 0xFF);
        }
        else
        {
            var header = BuildTlsRecordHeader(ApplicationData, payloadLen);
            Buffer.BlockCopy(header, 0, packet, 0, HeaderSize);
        }
        Buffer.BlockCopy(nonce, 0, packet, _headerSize, NonceSize);
        Buffer.BlockCopy(ciphertext, 0, packet, _headerSize + NonceSize, ciphertext.Length);
        return packet;
    }

    // ── M6: counter-derived data-plane nonce (mirrors Rust packet.rs) ────────────
    /// <summary>Build the 96-bit wire nonce for <paramref name="counter"/> as
    /// PRP(seed(4) ‖ counter_be(8)). A balanced Feistel network is bijective for any round
    /// function, so distinct (seed,counter) inputs — counter is monotonic — always map to
    /// distinct nonces (no AEAD nonce reuse), while the on-wire value no longer increments by
    /// 1 (no visible-counter DPI tell). Replaces the previous random 96-bit nonce.</summary>
    private byte[] NonceForCounter(long counter) => PrpNonce(_noncePrpKey, RawNonce(_nonceSeed, counter));

    /// <summary>The pre-permutation nonce input: seed(4) ‖ counter big-endian(8). Split out of
    /// <see cref="NonceForCounter"/> so the whole derivation is checkable against the shared
    /// fixture (<c>conformance/prp-nonce.json</c>) without constructing a codec.</summary>
    internal static byte[] RawNonce(byte[] seed, long counter)
    {
        var raw = new byte[NonceSize];
        Buffer.BlockCopy(seed, 0, raw, 0, 4);
        raw[4] = (byte)((counter >> 56) & 0xFF);
        raw[5] = (byte)((counter >> 48) & 0xFF);
        raw[6] = (byte)((counter >> 40) & 0xFF);
        raw[7] = (byte)((counter >> 32) & 0xFF);
        raw[8] = (byte)((counter >> 24) & 0xFF);
        raw[9] = (byte)((counter >> 16) & 0xFF);
        raw[10] = (byte)((counter >> 8) & 0xFF);
        raw[11] = (byte)(counter & 0xFF);
        return raw;
    }

    /// <summary>96-bit balanced Feistel permutation, 4 rounds; round fn = SHA256(key‖round‖half)[..6].
    /// Byte-for-byte identical to Rust <c>packet.rs prp_nonce</c> (not required for interop — the peer
    /// reads the nonce straight off the wire — but kept identical for auditability).</summary>
    internal static byte[] PrpNonce(byte[] key, byte[] raw)
    {
        var l = new byte[6];
        var r = new byte[6];
        Buffer.BlockCopy(raw, 0, l, 0, 6);
        Buffer.BlockCopy(raw, 6, r, 0, 6);
        for (byte round = 0; round < 4; round++)
        {
            var f = PrpRound(key, round, r);
            var nr = new byte[6];
            for (int i = 0; i < 6; i++) nr[i] = (byte)(l[i] ^ f[i]);
            l = r;
            r = nr;
        }
        var outp = new byte[NonceSize];
        Buffer.BlockCopy(l, 0, outp, 0, 6);
        Buffer.BlockCopy(r, 0, outp, 6, 6);
        return outp;
    }

    private static byte[] PrpRound(byte[] key, byte round, byte[] half)
    {
        var input = new byte[key.Length + 1 + half.Length];
        Buffer.BlockCopy(key, 0, input, 0, key.Length);
        input[key.Length] = round;
        Buffer.BlockCopy(half, 0, input, key.Length + 1, half.Length);
        var d = SHA256.HashData(input);
        var outp = new byte[6];
        Buffer.BlockCopy(d, 0, outp, 0, 6);
        return outp;
    }

    public byte[] Decrypt(byte[] packet)
    {
        if (packet.Length < _headerSize + NonceSize + TagSize + CounterSize + 2)
            throw new PacketException($"Packet too short: {packet.Length}");

        int payloadLen;
        if (_raw)
        {
            payloadLen = ((packet[0] & 0xFF) << 8) | (packet[1] & 0xFF);
        }
        else
        {
            if (packet[0] != ApplicationData)
                throw new PacketException($"Wrong content type: {packet[0]}");
            // The legacy_record_version too, not just the content type. Every record we EMIT
            // carries 0x03 0x03, and a real TLS 1.3 peer emits nothing else on an established
            // connection — so accepting other bytes made the masking framing looser than the
            // thing it imitates, for no gain. (Audit 2026-08-03, P3.)
            if (packet[1] != 0x03 || packet[2] != 0x03)
                throw new PacketException($"Wrong record version: {packet[1]:x2} {packet[2]:x2}");
            payloadLen = ((packet[3] & 0xFF) << 8) | (packet[4] & 0xFF);
        }
        if (payloadLen > MaxRecordSize)
            throw new PacketException($"Packet too large: {payloadLen}");
        // Defensive bounds (parity with the Rust decoder): the declared length must
        // be large enough to hold nonce+tag+counter+pad_len and must consume the complete
        // supplied record. Otherwise truncation would throw a raw range exception, while
        // trailing bytes would remain outside authentication. (L3)
        if (payloadLen < NonceSize + TagSize + CounterSize + 2
            || _headerSize + payloadLen != packet.Length)
            throw new PacketException(
                $"Packet length mismatch: payloadLen={payloadLen}, have={packet.Length - _headerSize}");

        var nonce = packet[_headerSize..(_headerSize + NonceSize)];
        var ciphertext = packet[(_headerSize + NonceSize)..(_headerSize + payloadLen)];

        var plaintext = _cipher.Decrypt(ciphertext, nonce);
        if (plaintext.Length < CounterSize + 2)
            throw new PacketException($"Decrypted payload too short: {plaintext.Length}");

        long packetCounter =
            ((long)(plaintext[0] & 0xFF) << 56) | ((long)(plaintext[1] & 0xFF) << 48) |
            ((long)(plaintext[2] & 0xFF) << 40) | ((long)(plaintext[3] & 0xFF) << 32) |
            ((long)(plaintext[4] & 0xFF) << 24) | ((long)(plaintext[5] & 0xFF) << 16) |
            ((long)(plaintext[6] & 0xFF) << 8) | (long)(plaintext[7] & 0xFF);

        // Validate the padding BEFORE recording the counter, like the Rust and iOS clients.
        // A record that decrypts (so it is genuinely from the peer) but carries malformed
        // padding is a peer bug, not an attack — recording it first needlessly burned that
        // counter's slot in the replay window. (Audit 2026-07-30.)
        int paddingLen = ((plaintext[^2] & 0xFF) << 8) | (plaintext[^1] & 0xFF);
        if (CounterSize + paddingLen + 2 > plaintext.Length)
            throw new PacketException($"Invalid padding: {paddingLen}");

        if (!AcceptCounter(packetCounter))
            throw new PacketException($"Replay detected: counter {packetCounter} (window highest {_replayHighest})");

        int dataLen = plaintext.Length - CounterSize - 2 - paddingLen;
        return plaintext[CounterSize..(CounterSize + dataLen)];
    }
}

File: qeli-ios/QeliCore/Protocol/PacketCodec.swift
import Foundation
import Security
import CryptoKit
@testable import Qeli

final class PacketCodec: @unchecked Sendable {
    static let tlsHeaderSize = 5
    static let nonceSize = 12
    static let tagSize = 16
    static let counterSize = 8
    static let replayWindow = 2_048
    static let replayWords = replayWindow / 64
    static let applicationData: UInt8 = 0x17
    static let maxRecordSize = 16_384 + nonceSize + tagSize + counterSize + 256

    private let cipher: PacketCipher
    private let rawFraming: Bool
    private let lock = NSLock()
    private var paddingEnabled: Bool
    private var paddingMin: Int
    private var paddingMax: Int
    private var counter: UInt64 = 0
    private var replayHighest: UInt64?
    private var replayBits = Array(repeating: UInt64(0), count: replayWords)
    // M6: per-instance nonce seed (4B) + PRP key (32B). The nonce goes on the wire and the peer
    // never inverts the PRP (it reads the nonce off the wire), so these are local randomness that
    // need NOT match the peer's — they only make (seed‖counter) unique per key, which a monotonic
    // counter + fresh per-session key guarantee.
    private let nonceSeed: Data
    private let noncePrpKey: Data

    init(
        cipher: PacketCipher,
        paddingEnabled: Bool = true,
        paddingMin: Int = 0,
        paddingMax: Int = 255,
        rawFraming: Bool = false
    ) {
        self.cipher = cipher
        self.paddingEnabled = paddingEnabled
        self.paddingMin = paddingMin
        self.paddingMax = paddingMax
        self.rawFraming = rawFraming
        self.nonceSeed = Self.secureRandom(4)
        self.noncePrpKey = Self.secureRandom(32)
    }

    var headerSize: Int { rawFraming ? 2 : Self.tlsHeaderSize }

    func setPadding(enabled: Bool, minimum: Int, maximum: Int) {
        lock.withLock {
            paddingEnabled = enabled
            paddingMin = minimum
            paddingMax = maximum
        }
    }

    func encrypt(_ plaintext: Data) throws -> Data {
        let padding = lock.withLock { () -> Int in
            guard paddingEnabled else { return 0 }
            let low = min(max(paddingMin, 0), 65_535)
            let high = min(max(paddingMax, low), 65_535)
            return high > low ? Int.random(in: low...high) : low
        }
        return try encrypt(plaintext, explicitPadding: padding)
    }

    func encryptCapped(_ plaintext: Data, maxInnerAndPadding: Int) throws -> Data {
        let padding = lock.withLock { () -> Int in
            guard paddingEnabled else { return 0 }
            let room = max(0, maxInnerAndPadding - plaintext.count)
            let low = min(max(paddingMin, 0), room)
            let high = min(max(paddingMax, low), room)
            return high > low ? Int.random(in: low...high) : low
        }
        return try encrypt(plaintext, explicitPadding: padding)
    }

    func encrypt(_ plaintext: Data, explicitPadding: Int) throws -> Data {
        let sequence = try lock.withLock { () throws -> UInt64 in
            guard counter < UInt64(Int64.max - 1_000) else { throw PacketCodecError.counterExhausted }
            defer { counter += 1 }
            return counter
        }
        let paddingLength = min(max(explicitPadding, 0), 65_535)
        // Counter-derived, collision-free, DPI-opaque nonce (was a random 96-bit value, which
        // carries a birthday-bound collision risk the Rust core eliminates). (client-audit M6)
        let nonce = nonceFor(sequence)
        let padding = try Self.randomData(count: paddingLength)
        var inner = Data()
        inner.reserveCapacity(Self.counterSize + plaintext.count + paddingLength + 2)
        inner.appendBigEndian(sequence)
        inner.append(plaintext)
        inner.append(padding)
        inner.append(UInt8((paddingLength >> 8) & 0xff))
        inner.append(UInt8(paddingLength & 0xff))

        let encrypted = try cipher.encrypt(inner, nonce: nonce)
        let payloadLength = nonce.count + encrypted.count
        guard payloadLength <= Self.maxRecordSize, payloadLength <= 65_535 else {
            throw PacketCodecError.recordTooLarge(payloadLength)
        }
        var record = Data()
        record.reserveCapacity(headerSize + payloadLength)
        if rawFraming {
            record.append(UInt8((payloadLength >> 8) & 0xff))
            record.append(UInt8(payloadLength & 0xff))
        } else {
            record.append(contentsOf: [
                Self.applicationData, 0x03, 0x03,
                UInt8((payloadLength >> 8) & 0xff), UInt8(payloadLength & 0xff)
            ])
        }
        record.append(nonce)
        record.append(encrypted)
        return record
    }

    func decrypt(_ packet: Data) throws -> Data {
        let minimum = headerSize + Self.nonceSize + Self.tagSize + Self.counterSize + 2
        guard packet.count >= minimum else { throw PacketCodecError.packetTooShort(packet.count) }
        if !rawFraming, packet[0] != Self.applicationData {
            throw PacketCodecError.wrongContentType(packet[0])
        }
        // The legacy_record_version too, not just the content type. Every record we EMIT
        // carries 0x03 0x03, and a real TLS 1.3 peer emits nothing else on an established
        // connection — so accepting other bytes made the masking framing looser than the thing
        // it imitates, for no gain. (Audit 2026-08-03, P3.)
        if !rawFraming, packet[1] != 0x03 || packet[2] != 0x03 {
            throw PacketCodecError.wrongContentType(packet[1])
        }
        let payloadLength = rawFraming
            ? (Int(packet[0]) << 8) | Int(packet[1])
            : (Int(packet[3]) << 8) | Int(packet[4])
        guard payloadLength <= Self.maxRecordSize else { throw PacketCodecError.recordTooLarge(payloadLength) }
        guard payloadLength >= Self.nonceSize + Self.tagSize + Self.counterSize + 2,
              headerSize + payloadLength == packet.count else {
            throw PacketCodecError.recordLengthMismatch(
                declared: payloadLength,
                available: packet.count - headerSize
            )
        }
        let nonceRange = headerSize..<(headerSize + Self.nonceSize)
        let encryptedRange = (headerSize + Self.nonceSize)..<(headerSize + payloadLength)
        let decrypted = try cipher.decrypt(packet[encryptedRange], nonce: packet[nonceRange])
        guard decrypted.count >= Self.counterSize + 2 else { throw PacketCodecError.truncatedPlaintext }

        let sequence = decrypted.prefix(Self.counterSize).reduce(UInt64(0)) { ($0 << 8) | UInt64($1) }
        let paddingLength = (Int(decrypted[decrypted.count - 2]) << 8) | Int(decrypted[decrypted.count - 1])
        guard Self.counterSize + paddingLength + 2 <= decrypted.count else {
            throw PacketCodecError.invalidPadding(paddingLength)
        }
        // Authenticate and validate the complete record before mutating replay state.
        // Otherwise a malformed packet could consume a fresh sequence number and make
        // a later canonical packet with that sequence look like a replay.
        try lock.withLock {
            guard acceptCounter(sequence) else { throw PacketCodecError.replay(sequence) }
        }
        let dataEnd = decrypted.count - paddingLength - 2
        return decrypted[Self.counterSize..<dataEnd]
    }

    /// True if `sequence` is fresh (not a replay / not too old); records it as seen.
    /// `internal` so the shared replay-window fixture (`conformance/replay-window.json`)
    /// can drive it directly — the window is pure state, and going through `decrypt`
    /// would need a valid record per sequence number.
    func acceptCounter(_ sequence: UInt64) -> Bool {
        guard let highest = replayHighest else {
            replayHighest = sequence
            replayBits[0] = 1
            return true
        }
        if sequence > highest {
            let advance = sequence - highest
            if advance >= UInt64(Self.replayWindow) {
                replayBits = Array(repeating: 0, count: Self.replayWords)
            } else {
                shiftWindow(Int(advance))
            }
            replayHighest = sequence
            replayBits[0] |= 1
            return true
        }
        let distance = highest - sequence
        guard distance < UInt64(Self.replayWindow) else { return false }
        let word = Int(distance / 64)
        let mask = UInt64(1) << UInt64(distance % 64)
        guard replayBits[word] & mask == 0 else { return false }
        replayBits[word] |= mask
        return true
    }

    private func shiftWindow(_ bits: Int) {
        let words = bits / 64
        let offset = bits % 64
        if offset == 0 {
            for index in stride(from: Self.replayWords - 1, through: 0, by: -1) {
                replayBits[index] = index >= words ? replayBits[index - words] : 0
            }
        } else {
            for index in stride(from: Self.replayWords - 1, through: 0, by: -1) {
                let low = index >= words ? replayBits[index - words] << UInt64(offset) : 0
                let high = index > words ? replayBits[index - words - 1] >> UInt64(64 - offset) : 0
                replayBits[index] = low | high
            }
        }
    }

    // ── M6: counter-derived data-plane nonce (mirrors Rust packet.rs / C# PacketCodec) ──
    /// Build the 96-bit wire nonce for `counter` as PRP(seed(4) ‖ counter_be(8)). A balanced
    /// Feistel network is bijective for any round function, so distinct (seed,counter) inputs —
    /// counter is monotonic — always map to distinct nonces (no AEAD reuse), while the on-wire
    /// value no longer increments by 1 (no visible-counter DPI tell).
    private func nonceFor(_ counter: UInt64) -> Data {
        Self.prpNonce(key: noncePrpKey, raw: Self.rawNonce(seed: nonceSeed, counter: counter))
    }

    /// The pre-permutation nonce input: seed(4) ‖ counter big-endian(8). Split out of
    /// `nonceFor` so the whole derivation is checkable against the shared fixture
    /// (`conformance/prp-nonce.json`) without constructing a codec.
    static func rawNonce(seed: Data, counter: UInt64) -> Data {
        var raw = Data()
        raw.reserveCapacity(nonceSize)
        raw.append(seed)               // 4 bytes
        raw.appendBigEndian(counter)   // 8 bytes
        return raw
    }

    /// 96-bit balanced Feistel permutation, 4 rounds; round fn = SHA256(key‖round‖half)[..6].
    /// Byte-for-byte identical to Rust `packet.rs prp_nonce` (not required for interop — the peer
    /// reads the nonce straight off the wire — but kept identical for auditability).
    static func prpNonce(key: Data, raw: Data) -> Data {
        let bytes = [UInt8](raw)
        var l = Array(bytes[0..<6])
        var r = Array(bytes[6..<12])
        for round in UInt8(0)..<UInt8(4) {
            let f = prpRound(key: key, round: round, half: r)
            var nr = [UInt8](repeating: 0, count: 6)
            for i in 0..<6 { nr[i] = l[i] ^ f[i] }
            l = r
            r = nr
        }
        return Data(l + r)
    }

    private static func prpRound(key: Data, round: UInt8, half: [UInt8]) -> [UInt8] {
        var input = Data()
        input.append(key)
        input.append(round)
        input.append(contentsOf: half)
        let digest = SHA256.hash(data: input)
        return Array(digest.prefix(6))
    }

    /// Non-throwing secure random for the init-time nonce seed / PRP key. SecRandomCopyBytes
    /// effectively never fails on iOS; the fallback only guarantees distinct per-instance
    /// material (nonce uniqueness needs distinct seeds, not CSPRNG-grade for this field).
    private static func secureRandom(_ count: Int) -> Data {
        var data = Data(count: count)
        let status = data.withUnsafeMutableBytes {
            SecRandomCopyBytes(kSecRandomDefault, count, $0.baseAddress!)
        }
        if status != errSecSuccess {
            for i in 0..<count { data[i] = UInt8.random(in: 0...255) }
        }
        return data
    }

    private static func randomData(count: Int) throws -> Data {
        if count == 0 { return Data() }
        var data = Data(count: count)
        let status = data.withUnsafeMutableBytes {
            SecRandomCopyBytes(kSecRandomDefault, count, $0.baseAddress!)
        }
        guard status == errSecSuccess else { throw PacketCodecError.randomFailure(status) }
        return data
    }
}

enum PacketCodecError: LocalizedError {
    case counterExhausted
    case packetTooShort(Int)
    case wrongContentType(UInt8)
    case recordTooLarge(Int)
    case recordLengthMismatch(declared: Int, available: Int)
    case truncatedPlaintext
    case invalidPadding(Int)
    case replay(UInt64)
    case randomFailure(OSStatus)

    var errorDescription: String? {
        switch self {
        case .counterExhausted: return "Packet counter exhausted; reconnect required."
        case .packetTooShort(let count): return "Packet is too short (\(count) bytes)."
        case .wrongContentType(let value): return "Unexpected TLS content type \(value)."
        case .recordTooLarge(let count): return "Record payload is too large (\(count) bytes)."
        case .recordLengthMismatch(let declared, let available):
            return "Record length mismatch (declared \(declared), available \(available))."
        case .truncatedPlaintext: return "Decrypted packet is truncated."
        case .invalidPadding(let count): return "Invalid packet padding length \(count)."
        case .replay(let sequence): return "Replay detected for packet \(sequence)."
        case .randomFailure(let status): return "Secure random generator failed (\(status))."
        }
    }
}

private extension Data {
    mutating func appendBigEndian(_ value: UInt64) {
        var bigEndian = value.bigEndian
        Swift.withUnsafeBytes(of: &bigEndian) { append(contentsOf: $0) }
    }
}
